src/hlsBinaries.ts

```diff
@@ -746,32 +743,58 @@ async function getHLSesfromMetadata(context: ExtensionContext, logger: Logger):
     window.showErrorMessage('Could not get release metadata');
     return null;
   }
-  const plat = match(process.platform)
-    .with('darwin', (_) => 'Darwin')
-    .with('linux', (_) => 'Linux_UnknownLinux')
-    .with('win32', (_) => 'Windows')
-    .with('freebsd', (_) => 'FreeBSD')
+  const plat: Platform | null = match(process.platform)
+    .with('darwin', (_) => 'Darwin' as Platform)
+    .with('linux', (_) => 'Linux_UnknownLinux' as Platform)
+    .with('win32', (_) => 'Windows' as Platform)
+    .with('freebsd', (_) => 'FreeBSD' as Platform)
     .otherwise((_) => null);
   if (plat === null) {
     throw new Error(`Unknown platform ${process.platform}`);
   }
-  const arch = match(process.arch)
-    .with('arm', (_) => 'A_ARM')
-    .with('arm64', (_) => 'A_ARM64')
-    .with('x32', (_) => 'A_32')
-    .with('x64', (_) => 'A_64')
+  const arch: Arch | null = match(process.arch)
+    .with('arm', (_) => 'A_ARM' as Arch)
+    .with('arm64', (_) => 'A_ARM64' as Arch)
+    .with('x32', (_) => 'A_32' as Arch)
+    .with('x64', (_) => 'A_64' as Arch)
     .otherwise((_) => null);
   if (arch === null) {
     throw new Error(`Unknown architecture ${process.arch}`);
   }
+  logger.info(`Platform constants: ${plat}, ${arch}`);
+
+  return findSupportedHlsPerGhc(plat, arch, metadata, logger);
+}
 
-  const map: ReleaseMetadata = new Map(Object.entries(metadata));
+export type Platform = 'Darwin' | 'Linux_UnknownLinux' | 'Windows' | 'FreeBSD';
+
+export type Arch = 'A_ARM' | 'A_ARM64' | 'A_32' | 'A_64';
+
+/**
+ * Find all supported GHC versions per HLS version supported on the given
+ * platform and architecture.
+ * @param platform Platform of the host.
+ * @param arch Arch of the host.
+ * @param metadata HLS Metadata information.
+ * @param logger Logger.
+ * @returns Map from HLS version to GHC versions that are supported.
+ */
+export function findSupportedHlsPerGhc(
+  platform: Platform,
+  arch: Arch,
+  metadata: ReleaseMetadata,
+  logger: Logger
+): Map<string, string[]> {
   const newMap = new Map<string, string[]>();
-  map.forEach((value, key) => {
-    const value_ = new Map(Object.entries(value));
-    const archValues = new Map(Object.entries(value_.get(arch)));
-    const versions: string[] = archValues.get(plat) as string[];
-    newMap.set(key, versions);
+  metadata.forEach((supportedArch, hlsVersion) => {
+    const supportedOs = supportedArch.get(arch);
+    if (supportedOs) {
+      const ghcSupportedOnOs = supportedOs.get(platform);
+      if (ghcSupportedOnOs) {
+        logger.log(`HLS ${hlsVersion} compatible with GHC Versions: ${ghcSupportedOnOs}`);
+        newMap.set(hlsVersion, ghcSupportedOnOs);
+      }
+    }
```

Comment:
This is the only tangible change. Everything else should be only logging statements.